causal-analysis/TowardsCausalThinking.ipynb

```diff
@@ -0,0 +1,2117 @@
+   "metadata": {},
+   "source": [
+    "##### Store Gateway\n",
+    "\n",
+    "Reference: [Existing Thanos Alerts](https://github.com/thanos-io/thanos/blob/main/mixin/alerts/store.libsonnet)\n",
+    "\n",
+    "- ACM does not have all of these alerts today.\n",
+    "\n",
+    "\n",
+    "|Number|Alert|Meaning|\n",
+    "|---:|:---|:---|\n",
+    "|1|ThanosStoreGrpcErrorRate|Thanos Store is failing to handle gRPC requests.|\n",
+    "|2|ThanosStoreSeriesGateLatencyHigh|Thanos Store has high latency for store series gate requests.|\n",
+    "|3|ThanosStoreBucketHighOperationFailures|Thanos Store Bucket is failing to execute operations.|\n",
+    "|4|ThanosStoreObjstoreOperationLatencyHigh|Thanos Store is having high latency for bucket operations.|\n",
+    "\n",
+    "\n",
+    "Notice - all of these are excellent symptoms. It does not explain why these happened.\n",
+    "\n",
+    "__storegw-health-indicator is a vector__\n",
+    "\n",
+    "$\\mathbf{storegw-health-indicator} = \\begin{bmatrix} ThanosStoreGrpcErrorRate \\\\\\ ThanosStoreSeriesGateLatencyHigh \\\\\\ ThanosStoreBucketHighOperationFailures \\\\\\ ThanosStoreObjstoreOperationLatencyHigh   \\end{bmatrix}$\n"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "7b9b4934",
+   "metadata": {},
+   "source": [
+    "#### Analysis with Causal Model (aka metrics)\n",
+    "\n",
+    "In this case, we will assign metrics to each of the nodes - and will model relationships between the metrics as mentioned earlier. Based on this, the above causal graphs can be slightly change."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "597034c6",
+   "metadata": {},
+   "source": [
+    "##### Compactor\n",
+    "\n",
+    "A causal model looks like - \n",
+    "\n",
+    "<!--\n",
+    "$compactor-health = \\mathbf{f}(is-compactor-out-ofspace,is-there-a-bad-block)$\n",
+    "-->\n",
+    "$\\mathbf{compactor-health} = \\mathbf{f}(is-compactor-out-ofspace,is-there-a-bad-block)$\n",
+    "\n",
+    "How do we capture if there is a bad block - this may be hard and we may require some approximations perhaps?\n",
+    "A bad compactor is a train wreck waiting to happen. This can quickly snowball into lot more issues if not taken care off right away.\n",
+    "  "
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "18e19d13",
+   "metadata": {},
+   "source": [
+    "##### Receiver\n",
+    "\n",
+    "\n",
+    "A causal model looks like - \n",
+    "\n",
+    "$\\mathbf{receiver-health} = \\mathbf{f}(time-series, number-of-receiver-replicas, is-receiver-pv-out-of-space, can-receiver-reach-obj-store)$\n"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "cebd6387",
+   "metadata": {},
+   "source": [
+    "##### Querier\n",
+    "\n",
+    "\n",
+    "A causal model looks like - \n",
+    "\n",
+    "$\\mathbf{querier-health} = \\mathbf{f}(number-of-time-series-per-query, number-of-simaltaneous-queries, number-of-querier-replicas, receiver-health, query-cache-health)$\n"
```

Comment:
I would include number of samples per query or time range of query, as that dictates fanout, and number of postings that you have to read from tsdb. Also store-health.
```suggestion
    "$\\mathbf{querier-health} = \\mathbf{f}(number-of-time-series-per-query, number-of-samples-per=query, number-of-simultaneous-queries, number-of-querier-replicas, receiver-health, store-health, query-cache-health)$\n"
```